shotover/src/codec/opensearch.rs

```diff
@@ -0,0 +1,323 @@
+use super::{CodecBuilder, CodecReadError, CodecWriteError, Direction};
+use crate::frame::{opensearch::HttpHead, Frame, MessageType, OpenSearchFrame};
+use crate::message::{Encodable, Message, Messages};
+use anyhow::Result;
+use bytes::{Buf, BytesMut};
+use http::{header, HeaderName, HeaderValue, Request, Response};
+use std::sync::Arc;
+use tokio_util::codec::{Decoder, Encoder};
+
+#[derive(Clone)]
+pub struct OpenSearchCodecBuilder {
+    direction: Direction,
+}
+
+impl CodecBuilder for OpenSearchCodecBuilder {
+    type Decoder = OpenSearchDecoder;
+    type Encoder = OpenSearchEncoder;
+
+    fn new(direction: Direction) -> Self {
+        Self { direction }
+    }
+
+    fn build(&self) -> (OpenSearchDecoder, OpenSearchEncoder) {
+        (
+            OpenSearchDecoder::new(self.direction),
+            OpenSearchEncoder::new(self.direction),
+        )
+    }
+
+    fn websocket_subprotocol(&self) -> &'static str {
+        "opensearch" //TODO
+    }
+}
+
+pub struct OpenSearchDecoder {
+    direction: Direction,
+    state: State,
+}
+
+impl OpenSearchDecoder {
+    pub fn new(direction: Direction) -> Self {
+        Self {
+            direction,
+            state: State::ParsingResponse,
+        }
+    }
+
+    fn decode_request(&self, src: &mut BytesMut) -> Result<Option<(usize, HttpHead, usize)>> {
+        let mut headers = [httparse::EMPTY_HEADER; 16];
+        let mut request = httparse::Request::new(&mut headers);
+
+        let amt = match request.parse(&src[0..src.len()]).unwrap() {
+            httparse::Status::Complete(amt) => amt,
+            httparse::Status::Partial => return Ok(None),
+        };
+        match request.version.unwrap() {
+            1 => (),
+            _version => panic!("error!"), // TODO version error
```

Comment:
considering this method returns a `Result`, all of these `TODO error handling` should just return an error instead of being left as a TODO.